[Security Solution][Notes] - note icon to all tables and timeline

PhilippeOberti opened the pull request:
## Summary

This PR continues to add functionality to the new notes feature for Security Solution. It focuses on adding a note icon to the alerts table.

- add a note icon to a few tables in the application:
  - alerts table on the alerts page
  - alerts table on the cases page
  - events table in the hosts page
  - events table in the network page
  - events table in the users page
  - all timeline screens, with the `unifiedComponentsInTimelineEnabled` on and off
- add a red dot the the notes icon in the tables mentioned above as well as the timeline tables, signifying when an document has notes associated with it

When the table is loaded, a call is made to fetch all the notes for the documents visible on that page.

### TODO

- [x] add similar logic for the events table
- [x] verify that things work correctly in timeline

https://github.com/elastic/kibana/assets/17276605/f7ce26ab-b344-464f-bca1-369b70c72503

### How to test

- make sure the feature flag is enabled in your `kibana.yml` file: `xpack.securitySolution.enableExperimental: ['securitySolutionNotesEnabled']`
- verify that the note icon appears on all the tables mentioned above as well as timeline
- verify that when creating a note for an alert or event, the note icon reflect that in all other places for the same document id
- verify that no changes should be visible if either `securitySolutionNotesEnabled` or `expandableFlyoutDisabled` are on

_Note: as the server side code isn't ready, the frontend is mocking the fectch notes call and always returns one note per document id. If you load the alerts page in this PR, it is expected to see one note for each row of the table._

![Screenshot 2024-06-26 at 10 36 53 AM](https://github.com/elastic/kibana/assets/17276605/a4b25b4a-36c5-48f3-a2c1-d40f751c6e6f)

https://github.com/elastic/security-team/issues/9375

### Checklist

- [x] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/packages/kbn-i18n/README.md)
- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios

PhilippeOberti committed:
[Security Solution][Notes] - note icon to alerts table

elasticmachine commented:
Pinging @elastic/security-threat-hunting-investigations (Team:Threat Hunting:Investigations)

PhilippeOberti commented:
buildkite test this

PhilippeOberti commented:
/ci

christineweng commented:
The UI in general LGTM. Noticed 2 odd behaviors while testing (could be because of the fake data): 

Whenever I click on the notes icon, or toggle the notes tab in flyout, it automatically added a note.


https://github.com/elastic/kibana/assets/18648970/63ef29dd-a92e-4650-a4f5-549a606b1ed8

When I try to delete a note, an error toast appears (which is fine), when I reopen the notes tab, the error toast appears again, it also persists when opening a different note.


https://github.com/elastic/kibana/assets/18648970/9dd405ca-b737-460f-a6b2-1a1c4ebc9da6

christineweng reviewed (commented):
The UI in general LGTM. Noticed 2 odd behaviors while testing (could be because of the fake data): 

Whenever I click on the notes icon, or toggle the notes tab in flyout, it automatically adds a note.


https://github.com/elastic/kibana/assets/18648970/63ef29dd-a92e-4650-a4f5-549a606b1ed8

Deleting a note led to an error toast appear (which is correct), but when reopening the notes tab, the error toast appears again, it also persists when opening a different note. 


https://github.com/elastic/kibana/assets/18648970/9dd405ca-b737-460f-a6b2-1a1c4ebc9da6

PhilippeOberti committed:
add the notes icons to timeline when using the unifiedComponentsInTimelineEnabled feature flag
also fetch notes for the events table used on the Explore users, hosts and network pages

PhilippeOberti commented:
> The UI in general LGTM. Noticed 2 odd behaviors while testing (could be because of the fake data):
> 
> Whenever I click on the notes icon, or toggle the notes tab in flyout, it automatically adds a note.
> 
> Deleting a note led to an error toast appear (which is correct), but when reopening the notes tab, the error toast appears again, it also persists when opening a different note.
> 

Hey @christineweng, thanks for testing the PR. These 2 behavior are expected. At the moment the server side code to fetch notes isn't merged, so I'm mocking the fetch calls and return one note per document id. If you change tab and come back, we fetch again, to make sure we always have the most up to date data, so we get another mocked note...
As for the delete, it is also expected, as the delete api does exist so I'm using it, but because the frontend has mocked data, the id of the note you're trying to delete does not actually exist, hence the error.

All of these will be very shortly fixed, when we merge @kqualters-elastic's server side [PR](https://github.com/elastic/kibana/pull/186946)

kqualters-elastic commented on x-pack/plugins/security_solution/public/common/components/header_actions/actions.tsx:
```diff
@@ -272,6 +282,29 @@ const ActionsComponent: React.FC<ActionProps> = ({
             />
           </>
         )}
+        {(!securitySolutionNotesEnabled || expandableFlyoutDisabled) &&
```
nit: with this condition (4 clauses) and the one above (3) pull them both out into variables that are named to indicate to a reader what the sum of these flags equates to.

kqualters-elastic reviewed (approved):
looks good to me! 👍

kqualters-elastic commented on x-pack/plugins/triggers_actions_ui/public/application/sections/alerts_table/alerts_table_state.tsx:
```diff
@@ -78,7 +78,7 @@ export type AlertsTableStateProps = {
   browserFields?: BrowserFields;
   onUpdate?: (args: TableUpdateHandlerArgs) => void;
-  onLoaded?: () => void;
+  onLoaded?: (alerts: Alerts) => void;
```
maybe change this to `onLoaded?: (alerts?: Alerts) => void;` . I'm surprised this is passing ci, since this says 1 argument is required, yet in observability solution it's used with 0.

cnasikas replied:
Indeed is surprising. I would say better to leave it as you have it because it will always pass an array of alerts (even empty).

logeekal replied:
@PhilippeOberti , Can't we use existing prop of `onUpdate` because it looks to be doing the same function, I guess? What do you think?

logeekal replied:
Additionally, I would suggest you to consider making the changes in below code instead of changing the Alert Table API. 

This is the hook to render the row actions such as `notes` in the Alert Table. We already have access to all the events here. which can be used to fetch the documents.

https://github.com/elastic/kibana/blob/aca2fdfff64b25dba1b7856385695bd524b70cd7/x-pack/plugins/security_solution/public/detections/hooks/trigger_actions_alert_table/use_actions_column.tsx#L45-L63

logeekal reviewed (commented):
Still Reviewing. In the meantime, I have added one comment which might require more discussion.

PhilippeOberti commented:
> @PhilippeOberti , Can't we use existing prop of onUpdate because it looks to be doing the same function, I guess? What do you think?

`onUpdate` seems to be called multiple times - for example on page change - while `onLoaded` is called only once.

PhilippeOberti commented:
> Additionally, I would suggest you to consider making the changes in below code instead of changing the Alert Table API.
> 
> This is the hook to render the row actions such as notes in the Alert Table. We already have access to all the events here. which can be used to fetch the documents.

<img width="177" alt="Screenshot 2024-06-28 at 11 03 46 AM" src="https://github.com/elastic/kibana/assets/17276605/08fc9a81-8989-412c-b277-82455bf6ddb9">

doing a simple console log it seems that we would be making many calls, probably one per document id? The `onLoaded` has the advantage of providing all ids at once which would be more efficient

logeekal commented:
>doing a simple console log it seems that we would be making many calls, probably one per document id? The onLoaded has the advantage of providing all ids at once which would be more efficient

That makes sense becuase `getUseActionColumn` is executed for each row. Thanks for checking it out.

logeekal reviewed (approved):
Looks good. Thanks. I would have preffered if there was no mocked data in `main` but even I cannot think of good alternative for desk testing. Overall looks good.
